File: 3. Behavioral Design Patterns/Observer/Python/ObserverTemplate.py
# Observer Pattern – Python Template

from abc import ABC, abstractmethod

class Observer(ABC):
    @abstractmethod
    def update(self):
        pass

class Subject(ABC):
    @abstractmethod
    def attach(self, observer):
        pass
    
    @abstractmethod
    def detach(self, observer):
        pass
    
    @abstractmethod
    def notify(self):
        pass

class ConcreteSubject(Subject):
    def __init__(self):
        self.observers = []
    
    def attach(self, observer):
        self.observers.append(observer)
    
    def detach(self, observer):
        self.observers.remove(observer)
    
    def notify(self):
        for observer in self.observers:
            observer.update()
    
    def do_something(self):
        # change state
        self.notify()

class ConcreteObserver(Observer):
    def update(self):
        print("ConcreteObserver: Reacted to the event")

def main():
    subject = ConcreteSubject()
    observer = ConcreteObserver()
    subject.attach(observer)
    subject.do_something()

if __name__ == "__main__":
    main()

File: 3. Behavioral Design Patterns/Observer/Java/ObserverTemplate.java
// Observer Pattern – Java Template

import java.util.*;

interface Observer { void update(); }
interface Subject { void attach(Observer o); void detach(Observer o); void notifyObservers(); }

class ConcreteSubject implements Subject {
    private final List<Observer> observers = new ArrayList<>();
    public void attach(Observer o){ observers.add(o); }
    public void detach(Observer o){ observers.remove(o); }
    public void notifyObservers(){ for (Observer o : observers) o.update(); }
    public void doSomething(){ /* change state */ notifyObservers(); }
}

class ConcreteObserver implements Observer { public void update(){ /* react to change */ } }

public class ObserverTemplate { public static void main(String[] args){ ConcreteSubject s = new ConcreteSubject(); s.attach(new ConcreteObserver()); s.doSomething(); } }
